src/fiats/neural_network_s.F90

```diff
@@ -934,9 +937,6 @@
 
                   iteration: &
                   block
-
-                    real a(maxval(self%nodes_), input_layer:output_layer) ! Activations
-                    real z(size(b,1),size(b,2)), delta(size(b,1),size(b,2))
```

Comment:
I'm concerned the proposed transformation might lead to incorrect behavior.

All three of these variables need to be semantically "local" to each concurrent iteration, because each concurrent iteration independently references and defines (reads and writes) these temporary variables as part of their execution.

This version of the code is the one for compilers lacking the F18 `local` locality specifier we'd prefer to use, and we worked-around that missing feature by instead _deliberately_ declaring/allocating these local variables _inside_ the body of the `do concurrent` iteration. By declaring them within the scope of the `DO CONCURRENT` iteration block, this unambiguously ensures the storage for these three variables is private to each iteration (the semantic this code requires for correctness). 

F23 11.1.7.5: (emphasis added)
> A construct or statement entity of a construct or statement within the DO CONCURRENT construct has SHARED locality if it has the SAVE attribute. If it **does not have the SAVE attribute, is a different entity in each iteration, similar to LOCAL locality.**

By moving the declaration outside `do concurrent` as proposed here, these variables effectively gain "unspecified locality", and instead become subject to the following semantic rules in F23 11.1.7.5 (inherited from F08 8.1.6.7): 

> If a variable has unspecified locality,    
> • if it is referenced in an iteration it shall either be previously defined during that iteration, or shall not be defined or become undefined during any other iteration; if it is defined or becomes undefined by more than one iteration it becomes undefined when the loop terminates;    
> [...]
> NOTE 3     
> The restrictions on the statements in a DO CONCURRENT construct are designed to ensure there are no data dependencies between iterations of the loop.

This wording is admittedly less definitive, but I read this to imply that these variables _still_ need to behave as if each concurrent iteration has its own private copy, nearly identical to the LOCAL specifier (ignoring differences in variable contents at loop entry/exit that are irrelevant in this code). 

TL;DR: I don't understand why this transformation "helps" the compiler in question. In particular, **if this transformation causes the compiler to emit code where concurrent iterations all share the same copy of these three variables, that would break the correctness of this code.** This procedure semantically requires each concurrent iteration to have a private/local copy of these three variables.